parser/src/infix.rs

```diff
@@ -0,0 +1,420 @@
+}
+
+impl OperatorTable {
+    fn new<I>(ops: I) -> OperatorTable
+        where I: IntoIterator<Item = (&'static str, Operator)>,
+    {
+        OperatorTable {
+            operators: ops.into_iter().map(|(name, op)| (name.to_string(), op)).collect(),
+            default: Operator::new(9, Fixity::Left),
+        }
+    }
+}
+
+impl Default for OperatorTable {
+    fn default() -> OperatorTable {
+        OperatorTable::new(vec![
+            ("*", Operator::new(7, Fixity::Left)),
+            ("/", Operator::new(7, Fixity::Left)),
+            ("%", Operator::new(7, Fixity::Left)),
+            ("+", Operator::new(6, Fixity::Left)),
+            ("-", Operator::new(6, Fixity::Left)),
+            (":", Operator::new(5, Fixity::Right)),
+            ("++", Operator::new(5, Fixity::Right)),
+            ("&&", Operator::new(3, Fixity::Right)),
+            ("||", Operator::new(2, Fixity::Right)),
+            ("$", Operator::new(0, Fixity::Right)),
+            ("==", Operator::new(4, Fixity::Left)),
+            ("/=", Operator::new(4, Fixity::Left)),
+            ("<", Operator::new(4, Fixity::Left)),
+            (">", Operator::new(4, Fixity::Left)),
+            ("<=", Operator::new(4, Fixity::Left)),
+            (">=", Operator::new(4, Fixity::Left)),
+            // Hack for some library operators
+            ("<<", Operator::new(9, Fixity::Right)),
+            (">>", Operator::new(9, Fixity::Left)),
+            ("<|", Operator::new(0, Fixity::Right)),
+            ("|>", Operator::new(0, Fixity::Left)),
+        ])
+    }
+}
+
+impl<'a> Index<&'a str> for OperatorTable {
+    type Output = Operator;
+
+    fn index(&self, name: &'a str) -> &Operator {
+        self.operators.get(name).unwrap_or_else(|| {
+            if name.starts_with("#") {
+                &self[name[1..].trim_left_matches(char::is_alphanumeric)]
+            } else {
+                &self.default
+            }
+        })
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub enum Error<Id> {
+    ConflictingFixities((Id, Operator), (Id, Operator)),
+}
+
+/// Reconstruct the infix expression using the correct associativities
+/// and precedences.
+///
+/// Inspired by [`Language.Haskell.Infix`].
+///
+/// [`Language.Haskell.Infix`]: https://hackage.haskell.org/package/infix-0.1.1/docs/src/Language-Haskell-Infix.html
+pub fn reparse<Id>(expr: SpannedExpr<Id>,
+                   symbols: &IdentEnv<Ident = Id>,
+                   operators: &OperatorTable)
+                   -> Result<SpannedExpr<Id>, Error<Id>> {
+    use base::pos;
+
+    let make_op = |lhs: Box<SpannedExpr<Id>>, op, rhs: Box<SpannedExpr<Id>>| {
+        let span = pos::span(lhs.span.start, rhs.span.end);
+        Box::new(pos::spanned(span, Expr::Infix(lhs, op, rhs)))
+    };
+
+    let mut infixes = Infixes::new(expr);
+    let mut arg_stack = Vec::new();
+    let mut op_stack = Vec::new();
+
+    loop {
+        if arg_stack.len() == 2 && op_stack.len() == 1 {
```

Comment:
Did you mean to link to something more specific? That is the same link I posted (copied from infix.rs).